src/StreamJsonRpc/Reflection/MessageFormatterProgressTracker.cs

```diff
@@ -0,0 +1,239 @@
+            {
+                return objectType;
+            }
+
+            Type iProgressOfTType = null;
+            lock (IProgressOfTTypeMap)
+            {
+                if (!IProgressOfTTypeMap.TryGetValue(objectType, out iProgressOfTType))
+                {
+                    iProgressOfTType = objectType.GetTypeInfo().GetInterfaces().FirstOrDefault(i => i.IsConstructedGenericType && i.GetGenericTypeDefinition() == typeof(IProgress<>));
+                    IProgressOfTTypeMap.Add(objectType, iProgressOfTType);
+                }
+            }
+
+            return iProgressOfTType;
+        }
+
+        /// <summary>
+        /// Saves a <see cref="ProgressParamInformation"/> instance obtained from the given <see cref="object"/> into <see cref="progressMap"/>.
+        /// </summary>
+        /// <param name="value">The object which should implement <see cref="IProgress{T}"/> to create the <see cref="ProgressParamInformation"/> instance.</param>
+        /// <returns>The assigned <see cref="long"/> progres ID.</returns>
+        public long AddProgressObjectToMap(object value)
+        {
+            Requires.NotNull(value, nameof(value));
+
+            if (this.RequestIdBeingSerialized == null)
+            {
+                throw new NotSupportedException(Resources.ProgressObjectInResponseOrNotificationError);
+            }
+
+            lock (this.progressLock)
+            {
+                long progressId = this.NextProgressId++;
+                this.requestProgressMap.Add(this.RequestIdBeingSerialized.Value, progressId);
+
+                this.progressMap.Add(progressId, new ProgressParamInformation(value));
+
+                return progressId;
+            }
+        }
+
+        /// <summary>
+        /// Clears the objects in <see cref="requestProgressMap"/> and <see cref="progressMap"/> associated to the given request ID.
+        /// </summary>
+        /// <param name="requestId">The id of the request whose associated object need to be cleared.</param>
+        public void ClearProgressMaps(long requestId)
+        {
+            lock (this.progressLock)
+            {
+                object progressId;
+
+                if (this.requestProgressMap.TryGetValue(requestId, out progressId))
+                {
+                    this.requestProgressMap.Remove(requestId);
+                    this.progressMap.Remove(progressId);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets the <see cref="ProgressParamInformation"/> object associated with the given progress id.
+        /// </summary>
+        /// <param name="progressId">The key to obtain the <see cref="ProgressParamInformation"/> object from <see cref="progressMap"/>.</param>
+        /// <param name="valueType">Output parameter to store the obtained <see cref="ProgressParamInformation"/> object.</param>
+        public bool GetProgressTypeToReport(object progressId, out ProgressParamInformation valueType)
```

Comment:
>GetProgressTypeToReport [](start = 20, length = 23)

This returns more than the progress type. The naming convention for methods that return bool and have an out parameter is `TryGet...`. So how about naming this `TryGetProgressObject`?